File: projectBlueWater/src/main/java/com/lasthopesoftware/storage/directories/specs/FakePrivateDirectoryLookup.java
package com.lasthopesoftware.storage.directories.specs;

import com.annimon.stream.Stream;
import com.lasthopesoftware.storage.GetFreeSpace;
import com.lasthopesoftware.storage.directories.GetPrivateDirectories;
import com.namehillsoftware.handoff.promises.Promise;

import org.jetbrains.annotations.NotNull;

import java.io.File;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

public class FakePrivateDirectoryLookup implements GetPrivateDirectories, GetFreeSpace {

	private final List<File> files = new ArrayList<>();

	@NotNull
	@Override
	public Promise<Collection<File>> promisePrivateDrives() {
		return new Promise<>(files);
	}

	public void addDirectory(String filePath, long freeSpace) {
		files.add(new FreeSpaceFile(filePath, freeSpace));
	}

	@Override
	public long getFreeSpace(@NotNull File file) {
		return Stream.of(files).filter(f -> f.getPath().startsWith(file.getPath())).findFirst().mapToLong(File::getFreeSpace).orElse(0);
	}
}

File: projectBlueWater/src/test/java/com/lasthopesoftware/storage/directories/FakePrivateDirectoryLookup.kt
package com.lasthopesoftware.storage.directories

import com.lasthopesoftware.storage.GetFreeSpace
import com.namehillsoftware.handoff.promises.Promise
import java.io.File

class FakePrivateDirectoryLookup : GetPrivateDirectories, GetFreeSpace {
	private val files = ArrayList<FreeSpaceFile>()

	override fun promisePrivateDrives(): Promise<Collection<File>> {
		return Promise(files)
	}

	fun addDirectory(filePath: String, freeSpace: Long) {
		files.add(FreeSpaceFile(filePath, freeSpace))
	}

	override fun getFreeSpace(file: File): Long {
		var path = file.path
		val filePaths = files.map { it.path }
		while (!filePaths.contains(path)) {
			val pathSeparatorIndex = path.lastIndexOf('/')
			if (pathSeparatorIndex < 0) return 0
			path = path.substring(0, pathSeparatorIndex)
		}
		val matchingPath = path
		val freeSpaceFile = files.firstOrNull { it.path == matchingPath }
		return freeSpaceFile?.freeSpace ?: 0
	}
}
